File: csharp/variable-length-quantity/VariableLengthQuantity.cs
using System;
using System.Collections.Generic;

public static class VariableLengthQuantity
{
    public static uint[] Encode(uint[] numbers)
    {
        var result = new List<uint>();
        for (int i = 0; i < numbers.Length; ++i)
        {
            var intermediate = new List<uint>();
            uint v = numbers[i];
            uint lastEntry = (byte)(v & 0x7F); // get rid of continuation bit for last entry
            intermediate.Add(lastEntry);
            v >>= 7;
            while (v > 0)
            {
                uint nextEntry = (byte)(v | 0x80);
                intermediate.Add((byte)nextEntry);
                v >>= 7;
            }
            intermediate.Reverse(); // have to reverse this intermediate result to get the right order
            result.AddRange(intermediate);
        }
        return result.ToArray();
    }
    public static uint[] Decode(uint[] bytes)
    {
        if ((bytes[bytes.Length - 1] & 0x80) != 0) throw new InvalidOperationException("Last entry must have a 0 continuation bit");
        var result = new List<uint>();
        int i = 0;
        while (i < bytes.Length)
        {
            int count = 1;
            while ((bytes[i] & 0x80) != 0 ) // count the bytes that make up a number
            {
                bytes[i] -= 0x80;  // could also do another &= 0x7F here, but since I know that the continuation bit is set...
                count++;
                i++;
            }
            i++; // increment to the first byte of the next number
            int shift = 7*(count - 1);
            uint b = 0;
            do
            {
                b += bytes[i - count] << shift;  // go through the bytes from left to right and shift them by the right multiple of 7 bits
                shift -= 7;
                count--;
            } while (count > 0);
            result.Add(b);
        }
        return result.ToArray();
    }
    
}

File: elixir/variable-length-quantity/lib/variable_length_quantity.ex
defmodule VariableLengthQuantity do
  import Bitwise
  @doc """
  Encode integers into a bitstring of VLQ encoded bytes
  """
  @spec encode(integers :: [integer]) :: binary
  def encode(integers) do
    integers
    |> Enum.flat_map(fn number -> get_parts(number) |> set_continuation_bits() end)
    |> :binary.list_to_bin()
  end

  defp get_parts(n) do # the first entry in the list will be the last byte
    if n == 0, do: [], else: [n &&& 0x7F | get_parts(n >>> 7)]
  end

  defp set_continuation_bits(parts) do
    if parts |> Enum.empty?() do
      [0]
    else # set the continuation bit in all entries except the first, then reverse the list
      parts
      |> Enum.with_index()
      |> Enum.map(fn {el, ind} -> if ind != 0, do: el ||| 0x80, else: el end)
      |> Enum.reverse
    end
  end

  @doc """
  Decode a bitstring of VLQ encoded bytes into a series of integers
  """
  @spec decode(bytes :: binary) :: {:ok, [integer]} | {:error, String.t()}
  def decode(bytes) do
    bytelist = :binary.bin_to_list(bytes)
    cond do
      bytelist |> Enum.empty?                              -> {:error, "empty binary"}
      bytelist |> Enum.reverse |> hd |> number_continues?  -> {:error, "incomplete sequence"}
      true                                                 -> {:ok, split(bytelist) |> Enum.map(&get_num/1)}
    end
  end

  defp number_continues?(b), do: (b &&& 0x80) == 0x80 # checks the continuation bit

  defp split(bytelist) do
    heads = bytelist |> Enum.take_while(&number_continues?/1) # so the last byte is not yet attached
    case bytelist |> Enum.drop(heads |> Enum.count) do
      [last | rest] -> [ heads ++ [last] | split(rest) ] # construct a list of lists here, attach the last byte to its heads
      _ -> []
    end
  end

  defp get_num(bytelist) do
    bytelist
    |> Enum.reverse
    |> Enum.with_index
    |> Enum.map(fn {el, ind} -> (el &&& 0x7F) <<< 7*ind end) # push the bits to the left by a multiple of 7
    |> Enum.sum
  end
end
